packages/eslint-plugin/typings/ts-eslint.d.ts

```diff
@@ -397,8 +397,218 @@ declare module 'ts-eslint' {
 
   // This isn't the correct signature, but it makes it easier to do custom unions within reusable listneers
   // never will break someone's code unless they specifically type the function argument
-  type RuleFunction = (node: never) => void;
-  type RuleListener = Record<string, RuleFunction>;
+  type RuleFunction<T extends TSESTree.BaseNode = never> = (node: T) => void;
+
+  type CodePathFunction<T extends TSESTree.BaseNode = TSESTree.Node> = (
+    codePath: Rule.CodePath,
+    node: TSESTree.BaseNode
+  ) => void;
+
+  type CodePathSegmentFunction<T extends TSESTree.BaseNode = TSESTree.Node> = (
+    codePathSegment: Rule.CodePathSegment,
+    node: TSESTree.BaseNode
+  ) => void;
+
+  type CodePathSegmentLoopFunction<
+    T extends TSESTree.BaseNode = TSESTree.Node
+  > = (
+    fromCodePathSegment: Rule.CodePathSegment,
+    toCodePathSegment: Rule.CodePathSegment,
+    node: TSESTree.BaseNode
+  ) => void;
+
+  interface RuleListener {
+    [nodeSelector: string]:
+      | RuleFunction
+      | CodePathFunction
```

Comment:
This is weird that TypeScript is making me do this. The type for the index signature is actually just `RuleFunction`. Can we fix this?

Reply:
This is happening on purpose; TS enforces that your defined keys match up to the index signature.
Because you have the code path keys returning the CodePathFunction, you have to make the index signature match.

You can get around this with intersection types:
```TS
type Fn1 = () => number;
type Fn2 = () => string;
type Fn3 = () => boolean;

type X = {
    [k: string]: Fn1
} & {
    key: Fn2
} & {
    other: Fn3
}

const x: X = (null as any);
type index = typeof x['a'] // type === Fn1
type key = typeof x.key // type === Fn2
type other = typeof x.other // type === Fn3
```
[repl](https://www.typescriptlang.org/play/#src=type%20Fn1%20%3D%20()%20%3D%3E%20number%3B%0D%0Atype%20Fn2%20%3D%20()%20%3D%3E%20string%3B%0D%0Atype%20Fn3%20%3D%20()%20%3D%3E%20boolean%3B%0D%0A%0D%0Atype%20X%20%3D%20%7B%0D%0A%20%20%20%20%5Bk%3A%20string%5D%3A%20Fn1%0D%0A%7D%20%26%20%7B%0D%0A%20%20%20%20key%3A%20Fn2%0D%0A%7D%20%26%20%7B%0D%0A%20%20%20%20other%3A%20Fn3%0D%0A%7D%0D%0A%0D%0Aconst%20x%3A%20X%20%3D%20(null%20as%20any)%3B%0D%0Atype%20index%20%3D%20typeof%20x%5B'a'%5D%20%2F%2F%20type%20%3D%3D%3D%20Fn1%0D%0Atype%20key%20%3D%20typeof%20x.key%20%2F%2F%20type%20%3D%3D%3D%20Fn2%0D%0Atype%20other%20%3D%20typeof%20x.other%20%2F%2F%20type%20%3D%3D%3D%20Fn3%0D%0A%0D%0A)

Reply:
Thanks for explaining it! I don't think intersection can go around it:

```ts
interface Known {
    foo?: () => boolean;
}

interface Unknown {
    [k: string]: () => number;
}

type All = Known & Unknown;

let a: All = {
    foo: () => true
}

let b: All = {
    randomKey: () => 1
}
```

> [ts] Property `foo` is incompatible with index signature 

[repl](https://www.typescriptlang.org/play/#src=interface%20Known%20%7B%0D%0A%20%20%20%20foo%3F%3A%20()%20%3D%3E%20boolean%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Unknown%20%7B%0D%0A%20%20%20%20%5Bk%3A%20string%5D%3A%20()%20%3D%3E%20number%3B%0D%0A%7D%0D%0A%0D%0Atype%20All%20%3D%20Known%20%26%20Unknown%3B%0D%0A%0D%0Alet%20a%3A%20All%20%3D%20%7B%0D%0A%20%20%20%20foo%3A%20()%20%3D%3E%20true%0D%0A%7D%0D%0A%0D%0Alet%20b%3A%20All%20%3D%20%7B%0D%0A%20%20%20%20randomKey%3A%20()%20%3D%3E%201%0D%0A%7D)

Reply:
in this case you should use `|`
```ts
interface Known {
    foo?: () => boolean;
}

interface Unknown {
    [k: string]: () => number;
}

type All = Known | Unknown;

let a: All = {
    foo: () => true
}

let b: All = {
    randomKey: () => 1
}
```

Reply:
```ts
interface Known {
  foo?(c: string): boolean;
}

interface Unknown {
  [k: string]: (a: number, b: number) => number;
}

type All = Known | Unknown & Known;

let a: All = {
  foo(c) {
    return c === ''
  },
  randomKey(c, b) {
    return c + b
  }
};

let b: All = {
  randomKey: a => 1
};
```

you should try something like this

Reply:
That's pretty clever! I think it's the same as `Known | Unknown`, no? 

It's getting into weird types territory... with `All = Known | Unknown & Known` I can do this:

```ts
let c: All = {
    foo: () => true,
    bar: () => ''
}
```

Reply:
i have one more idea for this
```ts
interface Known {
  foo?(c: string): boolean;
}

type Unknown = {
  [k: string]: (a: number, b: number) => number;
};

type All = Known & Unknown;

let a: All = {
  foo(c) {
    return c === ''
  },
  aaaa(c, b) {
    return c + b
  },
  bar: (a) => ''
};

let b: All = {
  randomKey: a => 1
};
```

[repl](https://www.typescriptlang.org/play/#src=interface%20Known%20%7B%0D%0A%20%20foo%3F(c%3A%20string)%3A%20boolean%3B%0D%0A%7D%0D%0A%0D%0Atype%20Unknown%20%3D%20%7B%0D%0A%20%20%5BP%3A%20string%5D%3A%20(a%3A%20number%2C%20b%3A%20number)%20%3D%3E%20number%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20All%20%3D%20Known%20%26%20Unknown%3B%0D%0A%0D%0Alet%20a%3A%20All%20%3D%20%7B%0D%0A%20%20foo(c)%20%7B%0D%0A%20%20%20%20return%20c%20%3D%3D%3D%20''%0D%0A%20%20%7D%2C%0D%0A%20%20aaaa(c%2C%20b)%20%7B%0D%0A%20%20%20%20return%20c%20%2B%20b%0D%0A%20%20%7D%2C%0D%0A%20%20bar%3A%20(a)%20%3D%3E%20''%0D%0A%7D%3B%0D%0A%0D%0Alet%20b%3A%20All%20%3D%20%7B%0D%0A%20%20randomKey%3A%20a%20%3D%3E%201%0D%0A%7D%3B%0D%0A)

Reply:
Let's punt on it for now.
We can solve this with a refactor of `createRule` down the line when we want to use the codepath selectors.
I have some ideas on that but there's no need atm